Global Styles: Add ContrastChecker to the Background, Text, Links, and Colors screens

andrewserong opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Part of addressing #48055

This PR introduces the `ContrastChecker` component to the Colors, Background color, Text color, and Links color screens in global styles — but only for the root global styles (e.g. not global styles set at the block level). This is for simplicity sake for now, as I imagine it could be tricky to reliably do contrast checking at the block level where we might not absolutely know what colors are being displayed in that context.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

As reported in #48055, while there is currently a contrast warning when adjusting colors for an individual block in the post editor, when adjusting global styles and previewing how it looks across a site (particularly while using the Style Book), there currently isn't that kind of feedback.

The proposal here is to add it in for some places where it will hopefully be useful, and where it can be added (fairly) reliable. In this case, that's at the root level of global styles, for the colors, background color, link colors, and text colors screens.

Note: Heading and Buttons are not included yet — it could be interesting to look at seeing if we can have the `ContrastChecker` component accept arbitrary other color values / labels to use in a follow-up PR.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

* Add a `useContrastCheckerColors` hook to the `color-utils.js` file in global styles, to group together common logic for retrieving the global styles values for background color, text color, and link color.
* In the parent Colors screen, the Text color screen, the Background color screen, and the Link colors screen, display the ContrastChecker component if the screen is viewing the respective root global styles values. Note: for the Link colors screen, it only uses the default link color and not the hover color for the contrast checker values.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

1. Open global styles and go to the Colors screen and adjust the root global styles colors for background color so that it is fairly close to the Text color. You should see the contrast warning.
2. Adjust the text color so that there is high contrast, and the warning should disappear.
3. Adjust the link color so that there is low contrast between link and background color. The warning should appear again.
4. Contrast warnings should not appear if you adjust block-level color values in global styles.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

| Colors screen | Background color screen | Text color screen | Link color screen |
| --- | --- | --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/219274123-039151e2-9f40-4384-9686-8c15a6ceb181.png) | ![image](https://user-images.githubusercontent.com/14988353/219274219-378796ba-69b0-4116-833e-b7e34b098f7f.png) | ![image](https://user-images.githubusercontent.com/14988353/219274287-6fdd3421-8fb1-44b7-b11c-b95972f868c1.png) | ![image](https://user-images.githubusercontent.com/14988353/219274512-51d8cb13-c078-4c91-854e-a66cb8c8b5f8.png) |

## Screengrab

The below screengrab demonstrates the visibility of the contrast checker while the Style Book is open. The hope is that for users playing around with their site-wide color settings, that the introduced contrast checker will provide a useful warning.

https://user-images.githubusercontent.com/14988353/219275092-fd7bca57-5bb2-4fa6-a126-84b0ecd37a72.mp4

andrewserong committed:
Global Styles: Add ContrastChecker to the Background, Text, Links, and Colors screens for root global styles

andrewserong commented:
@youknowriad I've added you as a reviewer just as an FYI as I know you've been looking into consolidating some of these panels in the block editor, and I wasn't sure if these global styles lookups might add some extra complexity for when the color panels consolidation happens.

Also, just pinging some designers as an FYI since I know there are some designers with an interest in the contrast checker notice: @WordPress/gutenberg-design

jameskoster commented:
A couple of observations from a quick glance at the screenshots:

In the suggestions ("Try using a...") I wonder if it would be simpler to reference the relevant panel only? IE in the Text panel, instead of:

> Try using a brighter background color and/or a darker text color

Just say:

> Try using a darker text color

What do you think?

In the Colors panel, it's not clear where the contrast issue is. It could be any of Headings, Text, Links, or Buttons. Is there a way to contextualise that information?

apeatling commented:
> In the Colors panel, it's not clear where the contrast issue is. It could be any of Headings, Text, Links, or Buttons. Is there a way to contextualise that information?

If I recall correctly the contrast checker component is very limited in terms of how you can alter the wording. The same thoughts came up in https://github.com/WordPress/gutenberg/pull/47205. It would need some reworking of the component to support passing custom messaging.

apeatling commented:
> In the Colors panel, it's not clear where the contrast issue is. It could be any of Headings, Text, Links, or Buttons. Is there a way to contextualise that information?

I agree with this, it is not clear what the contrast checker is talking about. It could be the background and text color of the buttons, or the site itself. I think it would need to be really clear and specific about what it's talking about to add any useful information for the user and not confuse them further.

apeatling reviewed (commented):
This seems like a good opportunity to upgrade the contrast checker component. 

We have a couple of places now with multiple color combinations in a single view. The checker is really only designed for reporting on a single instance of background/link/text colors and breaks down when you have more than a single instance (or different elements).

I mentioned above, but another example of the limitation was in the navigation block: https://github.com/WordPress/gutenberg/issues/46637 where the menu and submenu colors produced what looked like duplicate messages -- but really just needed custom wording.

andrewserong commented:
Thanks for the feedback @apeatling and @jameskoster, much appreciated! I agree with all your notes here, and from the sounds of it I think Andy's right — we'll likely need to upgrade the contrast checker component for what we're attempting to do here.

My ideas at the moment for some pragmatic next steps:

* Remove ContrastChecker from the Colors panel for now, as it's harder in that context to figure out how to communicate where the contrast issue is happening.
* Look into updating the ContrastChecker component so that we can configure the message a little and/or pass in a custom label for the current text color to use for the comparison — essentially, so that if you're on the Text color panel, the warning it gives you is about changing the Text color only — or if you're on the Background color panel, it only warns you about the Background color.

I'll have a play around with the ContrastChecker component to see what's possible, but let me know if anyone feels strongly about how it should be implemented.

> We have a couple of places now with multiple color combinations in a single view. The checker is really only designed for reporting on a single instance of background/link/text colors and breaks down when you have more than a single instance (or different elements).

Yeah, I was just thinking that when we have Headings or Buttons as well, the messaging on the Colors panel could get quite complex... I'm wondering if in the shorter-term it's better to focus on just the individual Background color, Text color, and Link color screens 🤔. Still, it'd be nice to figure out how to allow messaging for more complex combinations... if not now, then perhaps in a follow-up PR.

jameskoster commented:
That sounds like a fair plan.

The Colors panel probably needs per-row validation. Perhaps we could do something like this:

<img width="268" alt="Screenshot 2023-02-17 at 10 11 54" src="https://user-images.githubusercontent.com/846565/219615985-2f71601c-7682-4a8c-8f24-df0d88a06b59.png">

The Background / Text situation is a bit tricky, but maybe they should be combined into a single 'Body' panel anyway. I agree this is something to explore separately.

apeatling commented:
>  I'm wondering if in the shorter-term it's better to focus on just the individual Background color, Text color, and Link color screens 🤔. Still, it'd be nice to figure out how to allow messaging for more complex combinations... if not now, then perhaps in a follow-up PR.

Yes agree starting with the individuals for the short term, but exploring James' idea above would be great as a follow up.

andrewserong committed:
Try splitting the message so that the suggestion is only for the currently visible control

andrewserong committed:
Fix the label being in the wrong order

andrewserong commented:
Update: I've done the following:

* Removed the `ContrastChecker` from the parent colors screen.
* Updated the `ContrastChecker` component to accept a `suggestions` object with a flag for each of the color properties to provide a suggestion for. We can use this to define which property we'd like the suggestion to be about.

In general, this appears to work kinda okay when a user has one of the values set to an extreme (bright or dark), but highlights an issue with how the contrast checker suggests fixes for the low contrast when values are somewhere in the middle. When the message combines background + text color (as on trunk), then the user is nudged to change _either_ the background or the text color. However, it's possible to get into a situation where the suggestion for only a single property is misleading.

In the below screengrab, I've set the text color to a medium grey and the link color to black. Then, on the Background screen, adjusting the colors results in a suggestion to use either a brighter or darker color for the background color, but the contrast checker is never happy. Even when it's happy with the contrast with the text color, then the contrast checking for the link color kicks in (when the background is dark), and it's still telling me to change the background color: 😱

https://user-images.githubusercontent.com/14988353/220005512-ef8c4624-0598-4268-a3af-68b39f017110.mp4

For me, this makes the message more confusing than if we kept the behaviour where the opposing color is also described. When both color properties are included in the message, I think it might be more likely for a user to be able to find a solution to the warning than if we only mention the currently selected property. So, personally, I'd probably lean toward something like:

* Still remove the `ContrastChecker` from the parent colors screen.
* On the individual Background, Text, and Link colors screens, retain the full message that nudges to change either Background color or Text/link color — while the opposing color isn't immediately available from the current view, in this context the user has already selected the property that they're adjusting, so (hopefully) they'll be able to navigate their way back to the parent level to find the control for the opposing color.

Even so, it isn't really ideal, so I'd be curious for feedback here — are any of the above options confidently a better nudge to a user than no warning at all? My main objective here is for us to land on a contrast checker warning that is useful to the user without it being a frustrating warning that can't be appeased 😅

Thanks again for the feedback and help! Happy to try out any other ideas, too.

joedolson commented:
I think what makes the most sense is for the contrast checker to recommend changing the color the user has most recently changed. If you change the background color and it's not accessible, suggest changing the backgrond color.  If you changed the link color, suggest changing the link color. That gives the most direct feedback if something has changed.

The user is still able to change any color, so this doesn't stop other solutions, but it seems like feedback that's targeted at the most recent action a user took makes the most sense.

apeatling commented:
I think we've underestimated this issue! :)

Reading your post above @andrewserong it's clear we are not fully considering the fact that it's a multidimensional check of background/text/link and this breaks when simplifying the messaging. As far as I can tell it also means that @jameskoster's individual row reporting screen grab above also wouldn't work as presented unless it spanned background/text/color?

My opinion is that this may only make sense on the top level where you can see all the colors in a single place and are able to understand the context and dynamic between them all.

andrewserong commented:
> The user is still able to change any color, so this doesn't stop other solutions, but it seems like feedback that's targeted at the most recent action a user took makes the most sense.

In principle, I quite like that idea, however in practice I think we'll likely still end up with messaging that is confusing to the user. For example, if the color combination is already too low contrast before a user begins to interact with the control, or if the contrast affects an element that isn't currently visible. This gets to the problem that Andy mentions about the multi-dimensional nature of the contrast checker — if you're adjusting background color, then it's comparing against both text and link color, so it's likely that the right fix for the user could be to adjust more than one color control rather than just the one that's currently selected.

This gets a little tricky to describe, but I think my concern at this stage is that a warning that gives a user the wrong suggestion (or a misleading one) seems potentially worse to me than no warning at all. I'll continue this thought a little more below in response to Andy's feedback 🙂

> I think we've underestimated this issue! :)

I very much agree! It's funny how some of these ones sound superficially straightforward (let's place the existing component in this new context) but are then particularly difficult to get right 😅 

> My opinion is that this may only make sense on the top level where you can see all the colors in a single place and are able to understand the context and dynamic between them all.

The more that I think about it, I tend to agree — it's difficult to provide messaging / warnings in one context that are about something in a completely different screen. From playing around with warning messages at either the Background color screen or Text color screen, each time the warning felt like it either wasn't immediately actionable, or didn't understand my current intent (i.e. I want to switch from a light theme to a dark theme, and it's interrupting me in the process).

One of the things I like about seeing if we can communicate the contrast issues from the top level Colors screen is that we could treat the messaging there sort of like a "dashboard status" of where the current color contrast sits overall.

Some potential ideas for next steps:

* At the individual color screens (Background, Text, Link, etc), we could explore flagging that there's a color contrast issue without giving a nudge as to what to do about it — but perhaps that's also not very useful 🤔
* Explore @jameskoster's ideas for the per-row flagging of contrast issues from this comment: https://github.com/WordPress/gutenberg/pull/48115#issuecomment-1434425948 — or, if there are other options for how we can display or contextualise the current contrast issue from that view

In terms of this PR, I currently feel that the ideal direction isn't quite clear enough, so I doubt it's a quick or simple change to get this PR viable. To put it a slightly different way: now that we've successfully established that this is a more complex issue, I'd like to attempt to distil what we've learnt so far, and then write a wrap-up comment over in the linked issue (#48055). I can then close out this PR, and we can explore other ideas in follow-ups. I imagine a follow-up would probably involve extracting the color contrast logic from the `ContrastChecker` component so that we can have custom messaging / display within global styles.

Let me know if anyone has other ideas / feedback to add, and I can close this PR out tomorrow and write up a wrap up comment. Thanks again for all the collaboration and discussion so far! 🙇

andrewserong commented:
Thanks for all the feedback on this experiment, everyone! I've added a wrap-up comment on the linked issue here: https://github.com/WordPress/gutenberg/issues/48055#issuecomment-1441004280. I'll close out this PR now.
